app/controllers/links_controller.rb

```diff
@@ -29,6 +32,13 @@ def external_url(type:, options: {})
 
   private
 
+  def validate_root_url
+    unless params[:root_url].nil?
+      root_uri = URI.parse(params[:root_url])
+      not_found unless TRUSTED_DOMAINS.include?(root_uri.host) || TRUSTED_DOMAINS.any? { |d| root_uri.host.end_with?(".#{d}") }
```

Comment:
Should we log a warning here?

Reply:
You mean "someone tried to access a forbidden `root_url`"?

Reply:
Exactly.